src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/Razor/RazorDynamicFileInfoProvider.TextChangesTextLoader.cs

```diff
@@ -0,0 +1,88 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.LanguageServer.LanguageServer;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.HostWorkspace.Razor;
+
+internal partial class RazorDynamicFileInfoProvider
+{
+    private sealed class TextChangesTextLoader(
+        TextDocument? document,
+        IEnumerable<TextChange> changes,
+        byte[] checksum,
+        SourceHashAlgorithm checksumAlgorithm,
+        string? encoding,
+        Uri razorUri) : TextLoader
+    {
+        public override async Task<TextAndVersion> LoadTextAndVersionAsync(LoadTextOptions options, CancellationToken cancellationToken)
+        {
+            if (document is null)
+            {
+                var text = SourceText.From("").WithChanges(changes);
+                return TextAndVersion.Create(text, VersionStamp.Default.GetNewerVersion());
+            }
+
+            var sourceText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+            // Validate the checksum information so the edits are known to be correct
+            if (IsSourceTextMatching(sourceText))
+            {
+                var version = await document.GetTextVersionAsync(cancellationToken).ConfigureAwait(false);
+                var newText = sourceText.WithChanges(changes);
+                return TextAndVersion.Create(newText, version.GetNewerVersion());
+            }
+
+            return await GetFullDocumentFromServerAsync(razorUri, cancellationToken).ConfigureAwait(false);
+        }
+
+        private bool IsSourceTextMatching(SourceText sourceText)
+        {
+            if (sourceText.ChecksumAlgorithm != checksumAlgorithm)
+            {
+                return false;
+            }
+
+            if (sourceText.Encoding?.WebName.Equals(encoding) == false)
```

Comment:
This feels redundant, in that the checksums wouldn't match if the encodings were different, but I don't know enough about it to be sure.

Reply:
I think of it as `Encoding = how to read bytes for content, ChecksumAlgorithm = how to read bytes to produce a checksum` 

They're explicitly different because some encodings allow for things like different preambles but if stored as a UTF-16 string will be the same (or displayed to a user I suppose). 

I should have written this as `sourceText.Encoding?.WebName == encoding` though and idk why I was weird about it